From 7163a2dacefef833314daea32dd064a7b7ac89a2 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Mon, 18 Aug 2025 17:35:35 +0800
Subject: [PATCH 230/236] mtk: hostapd: add dfs channel state update cmd

Add a DFS state update command for offchain command.
When CAC is triggered by an offchain command, the DFS state of
the previous channel operated by offchain will not be reset.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 hostapd/ctrl_iface.c         | 4 ++--
 src/ap/ap_drv_ops.c          | 3 ++-
 src/ap/ap_drv_ops.h          | 2 +-
 src/ap/dfs.c                 | 6 +++---
 src/drivers/driver.h         | 5 +++++
 src/drivers/driver_nl80211.c | 3 +++
 src/drivers/nl80211_copy.h   | 5 +++++
 7 files changed, 21 insertions(+), 7 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 9f1d5b742..0ac053d04 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -3024,7 +3024,7 @@ static int hostapd_ctrl_iface_chan_switch(struct hostapd_iface *iface,
 					    background_settings.freq_params.he_enabled,
 					    background_settings.freq_params.eht_enabled,
 					    background_settings.freq_params.sec_channel_offset,
-					    oper_chwidth, seg0, seg1, true);
+					    oper_chwidth, seg0, seg1, true, false);
 		if (ret) {
 			wpa_printf(MSG_ERROR, "Background radar start dfs cac failed, %d",
 				   ret);
@@ -5229,7 +5229,7 @@ hostapd_ctrl_iface_set_offchain(struct hostapd_data *hapd, char *cmd,
 				  iface->conf->ieee80211ax,
 				  iface->conf->ieee80211be,
 				  secondary_chan, chwidth,
-				  seg0, seg1, true)) {
+				  seg0, seg1, true, true)) {
 		wpa_printf(MSG_ERROR, "DFS failed to start CAC offchannel");
 		iface->radar_background.channel = -1;
 		return -1;
diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 66fb6d25d..317be0306 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1147,7 +1147,7 @@ int hostapd_start_dfs_cac(struct hostapd_iface *iface,
 			  int he_enabled, bool eht_enabled,
 			  int sec_channel_offset, int oper_chwidth,
 			  int center_segment0, int center_segment1,
-			  bool radar_background)
+			  bool radar_background, bool update_state)
 {
 	struct hostapd_data *hapd = iface->bss[0];
 	struct hostapd_freq_params data;
@@ -1177,6 +1177,7 @@ int hostapd_start_dfs_cac(struct hostapd_iface *iface,
 		return -1;
 	}
 	data.radar_background = radar_background;
+	data.dfs_state_update = update_state;
 	data.link_id = -1;
 
 #ifdef CONFIG_IEEE80211BE
diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index 53320d38b..a4e06b529 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -151,7 +151,7 @@ int hostapd_start_dfs_cac(struct hostapd_iface *iface,
 			  int he_enabled, bool eht_enabled,
 			  int sec_channel_offset, int oper_chwidth,
 			  int center_segment0, int center_segment1,
-			  bool radar_background);
+			  bool radar_background, bool update_state);
 int hostapd_drv_do_acs(struct hostapd_data *hapd);
 int hostapd_drv_update_dh_ie(struct hostapd_data *hapd, const u8 *peer,
 			     u16 reason_code, const u8 *ie, size_t ielen);
diff --git a/src/ap/dfs.c b/src/ap/dfs.c
index 4af12b2b3..7795b2691 100644
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -968,7 +968,7 @@ int hostapd_handle_dfs(struct hostapd_iface *iface)
 		hostapd_get_oper_chwidth(iface->conf),
 		hostapd_get_oper_centr_freq_seg0_idx(iface->conf),
 		hostapd_get_oper_centr_freq_seg1_idx(iface->conf),
-		use_radar_background);
+		use_radar_background, false);
 
 	if (res) {
 		wpa_printf(MSG_ERROR, "DFS start_dfs_cac() failed, %d", res);
@@ -1031,7 +1031,7 @@ int hostapd_handle_dfs(struct hostapd_iface *iface)
 				      iface->conf->ieee80211ax,
 				      iface->conf->ieee80211be,
 				      sec, hostapd_get_oper_chwidth(iface->conf),
-				      cf1, cf2, true);
+				      cf1, cf2, true, false);
 
 		iface->radar_background.channel = channel->chan;
 		iface->radar_background.freq = channel->freq;
@@ -1204,7 +1204,7 @@ static void hostapd_dfs_update_background_chain(struct hostapd_iface *iface)
 				  iface->conf->ieee80211be,
 				  sec, hostapd_get_oper_chwidth(iface->conf),
 				  oper_centr_freq_seg0_idx,
-				  oper_centr_freq_seg1_idx, true)) {
+				  oper_centr_freq_seg1_idx, true, false)) {
 		wpa_printf(MSG_ERROR, "DFS failed to start CAC offchannel");
 		iface->radar_background.channel = -1;
 		return;
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 64839e6f0..f6b0e7688 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -958,6 +958,11 @@ struct hostapd_freq_params {
 	 * op_class: Operating class of the channel
 	 */
 	u8 op_class;
+
+	/**
+	 * dfs_state_update - Whether DFS state update is requested
+	 */
+	bool dfs_state_update;
 };
 
 /**
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index ee1a5ba82..a03a3be5b 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -5398,6 +5398,9 @@ static int nl80211_put_freq_params(struct nl_msg *msg,
 	if (freq->radar_background &&
 	    nla_put_flag(msg, NL80211_ATTR_RADAR_BACKGROUND))
 		return -ENOBUFS;
+	if (freq->dfs_state_update &&
+	    nla_put_flag(msg, NL80211_ATTR_DFS_STATE_UPDATE))
+		return -ENOBUFS;
 
 	return 0;
 }
diff --git a/src/drivers/nl80211_copy.h b/src/drivers/nl80211_copy.h
index 800f22ef3..d17c9493d 100644
--- a/src/drivers/nl80211_copy.h
+++ b/src/drivers/nl80211_copy.h
@@ -2960,6 +2960,9 @@ enum nl80211_commands {
  * @NL80211_ATTR_CRTI_UPDATE_EVENT: Type of critical update event for notification
  *	to userspace, contains a value of enum nl80211_crit_update_event (u8).
  *
+ * @NL80211_ATTR_DFS_STATE_UPDATE: Update the DFS state, especially when the radar
+ *	detection is triggered by the offchain command.
+ *
  * @NUM_NL80211_ATTR: total number of nl80211_attrs available
  * @NL80211_ATTR_MAX: highest attribute number currently defined
  * @__NL80211_ATTR_AFTER_LAST: internal use
@@ -3539,6 +3542,8 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_VENDOR_MTK_STA,
 
+	NL80211_ATTR_DFS_STATE_UPDATE,
+
 	__NL80211_ATTR_AFTER_LAST,
 	NUM_NL80211_ATTR = __NL80211_ATTR_AFTER_LAST,
 	NL80211_ATTR_MAX = __NL80211_ATTR_AFTER_LAST - 1
-- 
2.45.2

